WaveOut: Target a specific latency to ensure a consistent experience across different sample rates

sukibaby opened the pull request:
## Summary

With WaveOut, typically one defines the following values before starting a WaveOut thread:
- sample rate
- number of blocks
- size of each block, in frames

Based on these values, a particular latency will be calculated. StepMania historically follows this approach.

However, it is possible to target a specific latency instead, and extrapolate the needed number of blocks if the sample rate and block size values are known.

This PR refactors the driver to calculate the ideal number of blocks **based on a pre-determined optimal target latency and block size**. This allows us to guarantee a **nearly-imperceptible** difference in WaveOut latency (0.29 ms) between 44.1 kHz and 48 kHz.

## Theory

I've chosen an ideal value which was calculated to provide the smallest possible delta value (0.29 ms) between actual latency at both 44100 and 48000 Hz.

By comparison, there is a very significant delta in WaveOut latency between 44100 and 48000 Hz with the values leftover from SM5 or the current values. This difference would directly impact the player's global offset.

Further elaboration, proofs, and references are below:
<details>
<summary>Click to expand</summary>

When we consider how the WaveOut API works, we need to provide three of four values and the fourth is calculated. So, at 44100 Hz, with 15 blocks of 256 frames each, 15 * 256 gives us 3840 frames of latency (3840 / 44100 = ~87 ms). If we use this same configuration with 48000 Hz, we get a latency of 80ms which is a significant difference.

However, if we increase to 16 blocks at 48000 Hz, we get 4096 frames of latency (~85 ms), which is much closer to the 87 ms when using 15 blocks at 44100 Hz.

A target latency of 118 ms allows for the smallest possible delta when using a block size of 512 frames, and testing with sample rates of 44100 and 48000. At 44100 Hz / 512 frames / 118 ms target latency, we have a calculated latency of 127.7098 ms. At 48000 Hz / 512 frames/ 118 ms target latency, we have a calculated latency of 128 ms - a 0.29 ms difference.

I determined this by simulating the `CalculteNumBufferChunks` calculation for a range of 100 to 300.

For reference, here is the code from StepMania 5 / older ITGMania (pulled from v0.7.0):

```
const int channels = 2;
const int bytes_per_frame = channels*2;		/* 16-bit */
const int buffersize_frames = 1024*8;	/* in frames */
const int buffersize = buffersize_frames * bytes_per_frame; /* in bytes */

const int num_chunks = 8;
const int chunksize_frames = buffersize_frames / num_chunks;
const int chunksize = buffersize / num_chunks; /* in bytes */
```

</details>

## No functional changes

Nothing related to driver functionality is changed. We just change the way the buffer parameters are determined. This PR only changes code which runs before the WaveOut thread is launched.

Some stylistic changes are present in this PR due to the renaming of some incorrectly styled variables and the necessity of moving some variables to the header file.

sukibaby committed:
WaveOut: Target a specific latency to ensure a consistent experience across different sample rates

With WaveOut, typically one defines the following values before starting a WaveOut thread: sample rate, number of blocks, and size of each block in frames. Based on these values, a particular latency will be calculated. StepMania historically follows this approach.

However, it is possible to target a specific latency instead, and extrapolate the needed number of blocks if the sample rate and block size values are known.

This PR refactors the driver to calculate the ideal number of blocks based on a pre-determined optimal target latency and block size. This allows us to guarantee a nearly-imperceptible difference in WaveOut latency (0.29 ms) between 44.1 kHz and 48 kHz.

Nothing related to driver functionality is changed. We just change the way the buffer parameters are determined. This PR only changes code which runs before the WaveOut thread is launched.

When we consider how the WaveOut API works, we need to provide three of four values and the fourth is calculated. So, at 44100 Hz, with 15 blocks of 256 frames each, 15 * 256 gives us 3840 frames of latency (3840 / 44100 = ~87 ms). If we use this same configuration with 48000 Hz, we get a latency of 80ms which is a significant difference.

However, if we increase to 16 blocks at 48000 Hz, we get 4096 frames of latency (~85 ms), which is much closer to the 87 ms when using 15 blocks at 44100 Hz.

A target latency of 118 ms allows for the smallest possible delta when using a block size of 512 frames, and testing with sample rates of 44100 and 48000. At 44100 Hz / 512 frames / 118 ms target latency, we have a calculated latency of 127.7098 ms. At 48000 Hz / 512 frames/ 118 ms target latency, we have a calculated latency of 128 ms - a 0.29 ms difference.

I determined this by simulating the CalculteNumBufferChunks calculation for a range of 100 to 300.

teejusb commented on src/arch/Sound/RageSoundDriver_WaveOut.h:
```diff
@@ -28,11 +27,14 @@ class RageSoundDriver_WaveOut: public RageSoundDriver
 	bool m_bShutdown;
 	int m_iLastCursorPos;
 	bool b_InitSuccess;
+	int NUM_CHUNKS; // TODO rename wo_num_blocks
```
Can we fix all the namings in this PR? I don't see why they should be the way they currently are.

sukibaby replied:
Yes,  it was only to provide an easy-to-read diff.

teejusb commented on src/arch/Sound/RageSoundDriver_WaveOut.h:
```diff
@@ -17,22 +24,24 @@ class RageSoundDriver_WaveOut: public RageSoundDriver
-	bool m_bShutdown;
-	int m_iLastCursorPos;
-	bool b_InitSuccess;
+	WAVEHDR wo_buffers[kMaximumNumBlocks];
```
still reviewing but member variables should have trailing underscore

sukibaby replied:
ahhh ok

sukibaby replied:
pushing this change

teejusb commented on src/arch/Sound/RageSoundDriver_WaveOut.h:
```diff
@@ -9,6 +9,13 @@
+// of 118 ms, 13 buffers is sufficient for both 44100 and 48000 Hz sample
+// rates, but at least 20 buffers are needed if higher sample rates (such
+// as 96000 Hz) are used, or the game will crash during initialization.
+constexpr int kMaximumNumBlocks = 24;
```
I feel like it shouldn't hurt to just keep the size as 32 as before. Sometimes people do have some weirdly high sample rates from what I saw from the files my setup earlier.

sukibaby replied:
This is for the sample rate the game is running at, like the `SoundPreferredSampleRate=` option in preferences.  Not affected by the content's sample rate. Will revert to 32 regardless

teejusb commented on src/arch/Sound/RageSoundDriver_WaveOut.cpp:
```diff
@@ -103,38 +113,55 @@ int64_t RageSoundDriver_WaveOut::GetPosition() const
 }
 
 RageSoundDriver_WaveOut::RageSoundDriver_WaveOut()
+	: soundevent_handle(CreateEvent(nullptr, false, true, nullptr))
```
These seem out of order from how the members are declared. Good practice would be to preserve that order. Some compilers would give a warning about this.

sukibaby replied:
Resolved.

sukibaby committed:
WaveOut: correct variable naming, style updates

Adopt Google C++ style guide naming:

m_hWaveOut → wave_out_handle
m_hSoundEvent → sound_event_handle
m_aBuffers → buffers
m_iSampleRate → sample_rate
m_bShutdown → shutdown
m_iLastCursorPos → last_cursor_pos
b_InitSuccess → init_success
NUM_CHUNKS → num_blocks
BUFFERSIZE_FRAMES → num_frames_per_block
CHUNKSIZE → block_size

I've also fixed spacing in the constructor and reorganized the list of member variables in the header for easier reading, and added context to the comment in the header file, changed a static const int to constexpr int, etc

s

sukibaby committed:
WaveOut: remove hardcoded bits per sample value

To derive the bit depth, we can determine the number of bytes per frame (kChannels * 2), divide by the number of channels to get bytes per sample (2), and finally convert to bits by multiplying by 8 (16 bits).

sukibaby committed:
WaveOut: clamp the value of wo_num_blocks_

teejusb commented on src/arch/Sound/RageSoundDriver_WaveOut.cpp:
```diff
@@ -103,38 +113,55 @@ int64_t RageSoundDriver_WaveOut::GetPosition() const
+	, wo_frames_per_block_(0)
+	, wo_num_blocks_(1)
+	, wo_blocksize_(0)
+	, MixingThread()
```
Seems like MixingThread is first not last. I suppose might as well rename that variable as well while we're here.

sukibaby replied:
resolved

teejusb commented on src/arch/Sound/RageSoundDriver_WaveOut.h:
```diff
@@ -9,6 +9,13 @@
+// of 118 ms, 13 buffers is sufficient for both 44100 and 48000 Hz sample
+// rates, but at least 20 buffers are needed if higher sample rates (such
+// as 96000 Hz) are used, or the game will crash during initialization.
+constexpr int kMaximumNumBlocks = 32;
```
This can be part of the class. It doesn't need to be global (currently you're polluting the global namespace). It can be a private static constexpr int

sukibaby replied:
ok makes sense,  this value is a member `static const int` again
